readme
NAME
    DBIx::Web - Active Web Database Layer

DESCRIPTION
    This is a database layer for Perl scripts, implementing triggers, record
    versioning, access control at the record level, file attachments to
    records, materialised views. There is a web user interface also.

    Based on DBI and DB_File data engines. Based on CGI for user interface.

    Examples included:
    - Groupware Organizer and Notes,
      remake of CGI::Bus examples with CGI::Bus compatibility
    - Service Desk and CMDB application

AUTHOR
    Andrew V Makarow "makarow at mail.com", thanks to Anton Uemlyanin

COPYRIGHT
    This program is free software; you can redistribute it and/or modify 
    it under the same terms as Perl itself.

    Without specially written agreement prohibited to use completely,
    partially or in ideas for impairing nature industries in Archangel
    region (wood and timber, geological and mining, space and war); Russian
    procrastination structures and kgbists.

    A help in the revival of Pomor'e will be good: http://www.pomor-rus.org

PREREQUISITES
    Currently implemented and tested on Win32 Active Perl 5.6/5.8, Apache, IIS.

HOW TO BUILD AND INSTALL

    Type:
	
	perl Makefile.PL
	make
	make install
    Or:
	copy ./lib/DBIx/* into perl/site/lib/DBIx

    See examples, which should be configured and installed manually:
	- copy scripts to 'cgi-bin' directory.
	- edit 'script.cgi' source, set appropriate '-path', '-url', '-urf', 
	  another settings.
	- create '-icons' by downloading Apache icons if Microsoft IIS
	- start 'script.cgi -setup'. File store path and 
	  sample web server configuration will be created.
	- chmod o+rw,g+rw,a+rw '-path'
	- configure web server according to '-path' files.


history
# 2009-05-03 htmlStart() changed, conditional -expires=>'now'
# 2009-04-30 corrected cgiAction() switching to recKey() for 'recRead' and '-edit'
# 2009-04-02 changed -frmLsc in cgiQuery() to accept unmentioned fieldnames
# 2009-04-02 new -ovw option of htmlMQH()
# 2009-03-03 recWikn() may get table name as '-wikq'
# 2009-03-03 recWikn() corrected for deleted records
# 2009-03-02 cgiForm() corrected field hyperlink -form array with -version =>'-'
# 2009-03-02 cgiForm() extended  field hyperlink -form array with -wikn field name
# 2009-02-25 cgiList() corrected limiting ddlb width
# 2009-02-20 cmdb: -wikq corrected - used recRead(-version=>'-')
# 2009-02-20 recRead() corrected, -version=>'-' implemented
# 2009-02-03 htmlRFD() changed - checkbox moved from hyperlink
# 2009-01-10 starting 0.79 version
# 2008-12-10 ugroups() - $CACHE->{-ugroups}
# 2008-12-10 cmdb: cache a_hdesk_stbar()
# 2008-12-03 mistake: key:// hyperlinks valid only when -cgibus or table//key syntax
# 2008-12-01 cgiDDLB() changed: added 'option ... selected' for arrays and hashes
# 2008-12-01 cgiDDLB() changed: l.options.item(i).innerText !='' ? l.options.item(i).innerText.toLowerCase().indexOf(k)
# 2008-12-01 cmdb: reversed 'application' choice list
# 2008-12-01 new hreverse()
# 2008-11-28 cgiHelp(): added 'urlh',.. '#ftext'...
# 2008-11-26 cmdb-s: soft translation of 'object' and 'application' choices
# 2008-11-26 cmdb-s: update 'vtime/-rvcVerWhen', 'recprc' in old records
#		not useful because of old cgi available
# 2008-11-25 'sub{' -> 'sub {'
# 2008-11-18 cgiRun() changing, new -diero eval{} option for plex to DESTROY()
# 2008-11-18 #urole(role,name) substitution to SQL WHERE
# 2008-11-16 new dbiWsubst()/dbiWSft(): #ftext('str') substitution to SQL WHERE
# 2008-11-10 new diags() inspector - destructor count check and some data
# 2008-11-09 new htmlSubmitSpl(): '&hellip;'
# 2008-11-09 not needed '&mdash;', '&Chi;'
# 2008-11-09 PLEX 'DisableImpersonate': w32ufswtr() for tvdFTQuery() # w32IISdpsn()
# 2008-11-09 w32umail() slow - replace with umail()
# 2008-11-07 code inspection - recurse refs in 'sub{'/'sub {'
# 2008-11-05 constructor changed reusing object: new(class, self,-opt=>val)
# 2008-11-02 end() and DESTROY() edited
# 2008-11-02 ugfile() changed using Win32::Process::Create()
# 2008-10-30 cmdb: solutions association within tfvReferences()
# 2008-10-27 tfvReferences() changed: splicekeys() used; opts considered in hyperlink
# 2008-10-06 htmlMenu(): 'recNew'  excluded when record editing
# 2008-10-05 htmlStart(): -target=>'_blank' for $s->cgiHook('recFormRWQ') && $s->{-pcmd}->{-edit}
# 2008-10-04 recWikn(): new '-wikq'
# 2008-10-04 cgiDDLB($self,{field},'eqp' mode,{record}) became as documented, was cgiDDLB($s, $f, $em, $qm, $d)
# 2008-10-04 new w32ufswtr()
# 2008-10-03 cmdb: corrected '*subject' hyperlink
# 2008-10-03 cmdb: rename 'Process' -> 'Operation'; does not filled by default
# 2008-10-03 cmdb: classifier 'cmdbm'+'definition' options commented for 'Application', 'Location', 'Operation'
# 2008-10-03 cmdb: new 'definition' record type
# 2008-09-30 osCmd() corrections for 'xcacls','cacls', !win32
# 2008-09-30 osCmd('cacls') may buzz sometimes somewhy - 'xcacls' may be used
# 2008-09-21 starting 0.77 version
# 2008-09-17 new '-w32xcacls' switch
# 2008-09-05 corrected pthMk()
# 2008-09-01 redesig of pthCp()/pthStamp() - osCmd(xcopy) replace
# 2008-08-27 cgiRun() fix: if !defined($s->{-c}->{-search})
# 2008-06-23 starting 0.76 version
# 2008-06-04 recUpd(), recIns(): before rec read triggers: local $a->{-edit} =undef;
# 2008-06-04 cmdb.cgi - 'incident' states reduced
# 2008-06-04 cmdb.cgi - 'mrole' assign moved to -recFlim0R from -recEdt0R
# 2008-05-07 cmdb.cgi - 'mailto' editable
# 2008-05-01 corrected utf8enc(), utf8dec(); used 'Encode' module if $] >=5.008
# 2008-04-24 extended cptran() for $] >=5.008
# 2008-04-07 using '_search' frame / '-search' url
# 2008-04-07 chg cmdb hesdk 'mailto' =~/^(?:new|draft|appr-do|scheduled|do|delay|progress|rollback|ok-appr|no-appr|ok|no|deleted)$/
# 2008-04-07 chg dbiACLike(): replaced '\b' with '(?:^|,|;)\s*' and '(^|,|;)[:blank:]*'
# 2008-04-07 chg dbmSeek(): replaced '\b' with '(?:^|,|;)\s*'
# 2008-04-07 chg ugrexp(): replaced "=~/\\b('"  with "=~/(?:^|,|;)\\s*('"
# 2008-04-06 new {-c}->{-search}: -frame=>RIGHT distinguished
# 2008-03-28 new particular command {-cmd=>frmCall, -form=>'*.psp'} to run unregistered .psp
# 2008-03-26 new -search url param
# 2008-03-26 chg cgiHelp(): decorated <code> cgiHelp()
# 2008-03-25 new -search slot
# 2008-03-25 new -htmlMQH value in -pcmd by htmlMQH()
# 2008-03-25 display of future or periodical records - using -frmLsoAdd=>'Nowdays'
# 2008-03-25 ext webus: new 'Nowdays' view
# 2008-03-25 changed udispq() using $CACHE
# 2008-03-20 'Managers' -> 'Manager'; replace -frmLso with -urole
# 2008-03-20 chg cmdb.cgi: adding new -frmLso=>'Nowdays', deleting 'Today'
# 2008-03-20 chg cmdb.cgi: replace -frmLso with -urole
# 2008-03-20 chg cgiQuery(): new my $qq =$c->{-qwhere} && (($c->{-qwhere} =~/^(\[\[.*?\]\])/) || ($c->{-qwhere} =~/^(\/\*.*?\*\/)/)) && $1;
# 2008-03-20 chg cgiParse(): split 'frmLso' if ','
# 2008-03-17 testing htmlML() for _frmName.disabled
# 2008-03-17 improved htmlMB('back') for Mozilla
# 2008-03-17 corrected htmlML() switching on frameset
# 2008-03-17 new $LNG->{-die} and $LNG->{-warn} fixing \&CORE::die and \&CORE::warn undefined subroutine behaviour
# 2008-03-04 cgiHelp/htmlMenu/cgiParse improved
# 2008-03-03 starting 0.74 version
# 2008-02-29 extnd -help using html
# 2008-02-20 chgd cgiRun(), htmlML() restoring '_frame' lost needed for Mozilla
# 2008-02-20 chgd htmlML(), cgiParse() name postfixes: [+&.] -> [+&.^]
#		htmlML():	_frmName.value.match(/\\+\$/) -> (/[+^]\$/)
#		htmlML():	disabled auto '++' gen under 'frmCallNew'
#		htmlMChs():	disabled auto '++' gen under 'frmCallNew'
# 2008-02-19 new 'frmLsoff'

test.pl
#!perl -w
use strict;
use Test;
BEGIN { plan tests => 18 }

if (1) {
   print "\nRequired modules:\n";
   foreach my $m ('DBIx::Web', 'POSIX', 'Sys::Hostname', 'Fcntl', 'Symbol', 'IPC::Open2', 'File::Copy', 'Data::Dumper', 'Safe', 'CGI', 'CGI::Carp') {
     print "use $m\t";
     ok(eval("use $m; 'ok'"), 'ok');
   }
}

if (1) {
   print "\nOptional modules, dependent on features used:\n";
   foreach my $m ('Apache'
   , 'DBI'
   , 'DB_File') {
     print "use $m\t";
     skip(!eval("use $m; 1"), 1);
   }
}

if (1 && $^O eq 'MSWin32') {
   print "\nWin32 modules, dependent on features used:\n";
   foreach my $m ('Win32','Win32::API','Win32::TieRegistry') {
     print "use $m\t";
     ok(eval("use $m; 'ok'"), 'ok');
     # skip(!eval("use $m; 1"), 1);
   }
   foreach my $m ('cacls.exe') {
     print "$m\t";
     # ok(eval(!(`$m /?` && 1)), 'ok');
     skip(!(`$m /?` && 1), 1);
   }
}



examples/webus-sql.pl
#!perl -w
BEGIN {
print "CGI::Bus applications database installer/updater\n";
}
use DBI;

my $root =input('Database manager',    'root');
my $rpsw =input('Manager\'s password', '');
my $cgbd =input('Database name',       'cgibus');
my $cgbu =input('Database user',       'cgibus');
my $cgbp =input('User\'s password',    '********');

print "Connecting to 'DBI:mysql:mysql' as '$root'...\n";
my $mysql=DBI->connect("DBI:mysql:mysql",$root,$rpsw)
        ||die("Couls not connect to mysql as '$root'");
my $db   =$mysql;

print "Executing <DATA>, some SQL DML error messages may be ignored...\n\n";
my $row;
my $cmd ='';
my $cmt ='';
while ($row =<DATA>) {
  chomp($row);
  if ($cmd && ($row =~/^#/ || ($cmd !~/^\s*\{/ && $cmd =~/;\s*$/) )) {
     my $v;
     chomp($cmd);
     print $cmt ||$cmd, " -> ";
     if   ($cmd =~/^\s*\{/) {$v =eval($cmd);   print $@ ? $@ : 'ok'}
     else {$v =$db->do($cmd); print $db->err ? $db->errstr : 'ok'}
     print ': ', defined($v) ? $v : 'null', "\n\n";
     $cmd ='';
     $cmt ='';
  }
  next if $row =~/^\s*#*\s*$/;
  if    ($row =~/^#/ && $cmd !~/^\s*\{/) {
        $cmt =$row;
  }
  elsif ($row =~/^\s*#/ || $row eq '') {
  }
  else {
        $cmd .=($cmd ? "\n" : '') .$row;
  }  
}


sub input {
 my ($pr, $dv) =@_;
 print $pr, @_ >1 ? ' [' .(defined($dv) ? $dv : 'null') .']' :'', ': ';
 my $r =<STDIN>;
 chomp($r);
 $r eq '' ? $dv : $r
}

#
##########################################
# DATABASE DEFINITIONS & UPDATES
##########################################
#
__END__
#
#
CREATE DATABASE IF NOT EXISTS cgibus;
#
#
{$db->do("GRANT ALL PRIVILEGES ON ${cgbd}.* TO ${cgbu}\@localhost IDENTIFIED BY '${cgbp}' WITH GRANT OPTION;")}
#
#
{$db=undef; $db =DBI->connect("DBI:mysql:$cgbd",$cgbu,$cgbp); <STDIN>}
#
#
# Groupware Organizer PDM
 #========================
 # '-' - reserved fields
#
#
create table gworganizer (
        id       varchar(60) primary key,
        idnv     varchar(60),  #   new version (value) pointer
        idpr     varchar(60),  #   previous record pointer
        idrm     varchar(60),  # + reply master pointer
        idrr     varchar(60),  # + reply root pointer
        idpt     varchar(60),  # - point to record
        idlr     varchar(60),  # - location record pointer
        lslote   varchar(60),  # - location slot
        cuser    varchar(60),  #   creator user
        ctime    datetime,     #   created time
        uuser    varchar(60),  #   updator user
        utime    datetime,     #   updated time
        puser    varchar(60),  #   principal user
        prole    varchar(60),  #   principal role
        auser    varchar(60),  #   actor user
        arole    varchar(60),  #   actor role
        aopt     varchar(10),  # - actor options: 'respond', 'see', 'edit'
        rrole    varchar(60),  #   reader role
        mailto   varchar(255), # + mail receipients
        mailtime datetime,     # - mailed time
        status   varchar(10),  #   record status
        progress decimal,      # - progress of the work
        etime    datetime,     #   end time
        stime    datetime,     #   start time
        period   varchar(20),  # + period of record (y, m, d, h)
        record   varchar(10),  #   record type
        object   varchar(60),  #   object name
        doctype  varchar(60),  #   document type
        subject  varchar(255), #   subject, title
        comment  text          #   comment, text
) # TYPE = BDB  # use mysqld-max, do not use fulltext
;
#
 #CREATE INDEX idnv     ON gworganizer (idnv,    etime);
 #CREATE INDEX idpr     ON gworganizer (idpr,    etime);
 #CREATE INDEX idrm     ON gworganizer (idrm,    etime);
 #CREATE INDEX idrr     ON gworganizer (idrr,    etime);
 #CREATE INDEX record   ON gworganizer (record,  etime);
 #CREATE INDEX object   ON gworganizer (object,  etime);
 #CREATE INDEX doctype  ON gworganizer (doctype, etime);
 #CREATE INDEX auser    ON gworganizer (auser,   etime);
 #CREATE INDEX arole    ON gworganizer (arole,   etime);
 #CREATE FULLTEXT INDEX ftext ON gworganizer 
 #                    (object,doctype,subject,comment);
#
#
#
# Notes PDM (mysql)
 #========================
 # '-' - reserved fields
create table notes (
        id       varchar(60) primary key,
        idnv     varchar(60),  #   new version (value) pointer
        idpr     varchar(60),  #   previous record pointer
        idrm     varchar(60),  #   reply master pointer
        cuser    varchar(60),  #   creator user
        ctime    datetime,     #   created time
        uuser    varchar(60),  #   updator user
        utime    datetime,     #   updated time
        prole    varchar(60),  #   principal role
        rrole    varchar(60),  #   reader role
        status   varchar(10),  #   record status
        record   varchar(10),  # - record type
        object   varchar(60),  # - object name
        doctype  varchar(60),  # - document type
        subject  varchar(255), #   subject, title
        comment  text          #   comment, text
) # TYPE = BDB  # use mysqld-max, do not use fulltext
;
#
 #CREATE INDEX idnv     ON notes (idnv,    utime);
 #CREATE INDEX idrm     ON notes (idrm,    utime);
 #CREATE INDEX cuser    ON notes (cuser,   utime);
 #CREATE INDEX uuser    ON notes (uuser,   utime);
 #CREATE INDEX prole    ON notes (prole,   utime);
 #CREATE INDEX rrole    ON notes (rrole,   utime);
 #CREATE INDEX subject  ON notes (subject, utime);

 #CREATE FULLTEXT INDEX ftext ON notes 
 #                     (subject,comment);
#
#
#
{"2002-04-09 'gwo' update"
 ###########################
}
#
#
DROP   INDEX idnv     ON gworganizer;
CREATE INDEX idnv     ON gworganizer (idnv,    etime, utime);

 #DROP   INDEX idnv_st  ON gworganizer;
 #CREATE INDEX idnv_st  ON gworganizer (idnv,    status, etime);
 #DROP   INDEX status   ON gworganizer;
 #CREATE INDEX status   ON gworganizer (status,  idnv,   etime);

DROP   INDEX idpr     ON gworganizer;
CREATE INDEX idpr     ON gworganizer (idpr,    etime, utime);
DROP   INDEX idrm     ON gworganizer;
CREATE INDEX idrm     ON gworganizer (idrm,    etime, utime);
DROP   INDEX idrr     ON gworganizer;
CREATE INDEX idrr     ON gworganizer (idrr,    etime, utime);
DROP   INDEX record   ON gworganizer;
CREATE INDEX record   ON gworganizer (record,  etime, utime);
DROP   INDEX object   ON gworganizer;
CREATE INDEX object   ON gworganizer (object,  etime, utime);
DROP   INDEX doctype  ON gworganizer;
CREATE INDEX doctype  ON gworganizer (doctype, etime, utime);
DROP   INDEX auser    ON gworganizer;
CREATE INDEX auser    ON gworganizer (auser,   etime, utime);
DROP   INDEX arole    ON gworganizer;
CREATE INDEX arole    ON gworganizer (arole,   etime, utime);

 #DROP   INDEX uuser    ON gworganizer;
 #CREATE INDEX uuser    ON gworganizer (uuser,   etime, utime);
 #DROP   INDEX cuser    ON gworganizer;
 #CREATE INDEX cuser    ON gworganizer (cuser,   etime, utime);
 #DROP   INDEX rrole    ON gworganizer;
 #CREATE INDEX rrole    ON gworganizer (rrole,   etime, utime);

DROP   INDEX ftext    ON gworganizer;
 #CREATE FULLTEXT INDEX ftext ON gworganizer(object,doctype,subject,comment);
#
#
#
{"2002-04-09 'notes' update"
 ###########################
}
#
#
DROP   INDEX idnv     ON notes;
CREATE INDEX idnv     ON notes (idnv,    utime desc,  ctime desc);

 #DROP   INDEX idnv_st  ON notes;
 #CREATE INDEX idnv_st  ON notes (idnv,    status, utime desc, ctime desc);
 #DROP   INDEX utime    ON notes;
 #CREATE INDEX utime    ON notes (utime desc,  ctime desc);

DROP   INDEX idrm     ON notes;
CREATE INDEX idrm     ON notes (idrm,    utime desc,  ctime desc);
DROP   INDEX cuser    ON notes;
CREATE INDEX cuser    ON notes (cuser,   utime desc,  ctime desc);
DROP   INDEX uuser    ON notes;
CREATE INDEX uuser    ON notes (uuser,   utime desc,  ctime desc);
DROP   INDEX prole    ON notes;
CREATE INDEX prole    ON notes (prole,   utime desc,  ctime desc);
DROP   INDEX rrole    ON notes;
CREATE INDEX rrole    ON notes (rrole,   utime desc,  ctime desc);
DROP   INDEX subject  ON notes;
CREATE INDEX subject  ON notes (subject, utime desc,  ctime desc);

DROP   INDEX ftext    ON notes;
 #CREATE FULLTEXT INDEX ftext ON notes (subject, comment);
#
#
#
{"2002-05-19 'notes' update"
 ###########################
}
#
ALTER TABLE notes ADD COLUMN mailto varchar(255) AFTER rrole
#
#
#
{"2003-10-21 'gworganizer' update"
 ###########################
}
#
ALTER TABLE gworganizer ADD COLUMN project varchar(60) AFTER object;
ALTER TABLE gworganizer ADD COLUMN cost    decimal     AFTER project;
DROP   INDEX project   ON gworganizer;
CREATE INDEX project   ON gworganizer (project,  etime, utime);
#
#
#

examples/cmdb-sql.pl
#!perl -w
BEGIN {
print "DBIx::Web CMDB application installer/updater\n";
}
use DBI;

my $root =input('Database manager',    'root');
my $rpsw =input('Manager\'s password', '');
my $cgbd =input('Database name',       'cgibus');
my $cgbu =input('Database user',       'cgibus');
my $cgbp =input('User\'s password',    '********');

print "Connecting to 'DBI:mysql:mysql' as '$root'...\n";
my $mysql=DBI->connect("DBI:mysql:mysql",$root,$rpsw)
        ||die("Couls not connect to mysql as '$root'");
my $db   =$mysql;

print "Executing <DATA>, some SQL DML error messages may be ignored...\n\n";
my $row;
my $cmd ='';
my $cmt ='';
while ($row =<DATA>) {
  chomp($row);
  if ($cmd && ($row =~/^#/ || ($cmd !~/^\s*\{/ && $cmd =~/;\s*$/) )) {
     my $v;
     chomp($cmd);
     print $cmt ||$cmd, " -> ";
     if   ($cmd =~/^\s*\{/) {$v =eval($cmd);   print $@ ? $@ : 'ok'}
     else {$v =$db->do($cmd); print $db->err ? $db->errstr : 'ok'}
     print ': ', defined($v) ? $v : 'null', "\n\n";
     $cmd ='';
     $cmt ='';
  }
  next if $row =~/^\s*#*\s*$/;
  if    ($row =~/^#/ && $cmd !~/^\s*\{/) {
        $cmt =$row;
  }
  elsif ($row =~/^\s*#/ || $row eq '') {
  }
  else {
        $cmd .=($cmd ? "\n" : '') .$row;
  }  
}


sub input {
 my ($pr, $dv) =@_;
 print $pr, @_ >1 ? ' [' .(defined($dv) ? $dv : 'null') .']' :'', ': ';
 my $r =<STDIN>;
 chomp($r);
 $r eq '' ? $dv : $r
}

#
##########################################
# DATABASE DEFINITIONS & UPDATES
##########################################
#
__END__
#
#
CREATE DATABASE IF NOT EXISTS cgibus;
#
#
{$db->do("GRANT ALL PRIVILEGES ON ${cgbd}.* TO ${cgbu}\@localhost IDENTIFIED BY '${cgbp}' WITH GRANT OPTION;")}
#
#
{$db=undef; $db =DBI->connect("DBI:mysql:$cgbd",$cgbu,$cgbp); <STDIN>}
#
#
# CMDB PDM
 #========================
 # '-' - reserved fields
#
#
{"'cmdbm' table"
 ###########################
}
#
create table cmdbm (
	id	varchar(60) primary key,
	idold	varchar(60),	# migration ptr
	idnv	varchar(60),	# new version (value) pointer
	cuser	varchar(60),	# creator user
	ctime	datetime,	# created time
	uuser	varchar(60),	# updator user
	utime	datetime,	# updated time
	vtime	datetime,	# versioned time

	authors	varchar(60),	# actor role
	readers	varchar(60),	# reader role
	status	varchar(10),	# record status

	record  	varchar(20),	# record type
	name		varchar(80),	# record name
	definition	varchar(255),
	stmt		varchar(80),

	system		varchar(80),	# 
	service		varchar(80),	# 
	application	varchar(80),	# 

	type		varchar(80),	# 
	model		varchar(80),	# 
	invno		varchar(40),
	location	varchar(80),
	cpu		varchar(40),
	ram		decimal,
	hdd		decimal,
	hardware	varchar(80),
	os		varchar(80),

	action		varchar(80),	# 
	computer	varchar(80),	# 
	slot		varchar(80),	# 
	device		varchar(80),	# 
	port		varchar(80),	# 
	ipaddr		varchar(20),
	ipmask		varchar(20),
	macaddr		varchar(40),
	speed		varchar(10),
	duplex		varchar(10),

	interface	varchar(80),	# 

	ugroup		varchar(80),	# 
	role		varchar(80),	# 
	user		varchar(80),	# 
	userdef		varchar(255),	#
	office		varchar(80),	# 

	description	varchar(80),	# 

	comment		text		#   comment, text
)	# TYPE = BDB	# using mysqld-max
;
#
#
#
{"'cmdbm' indexes"
 ###########################
}
#
 DROP   INDEX name     ON cmdbm;
 CREATE INDEX name     ON cmdbm (name, idnv);
 DROP   INDEX idnv     ON cmdbm;
 CREATE INDEX idnv     ON cmdbm (idnv,    utime);
#
#
#
{"'hdesk' table"
 ###########################
}
# 
CREATE TABLE hdesk (
	id	varchar(60) NOT NULL	default '',
	idnv	varchar(60)	default NULL,
	idpr	varchar(60)	default NULL,
	idrm	varchar(60)	default NULL,
	idrr	varchar(60)	default NULL,
	cuser	varchar(60)	default NULL,
	ctime	datetime	default NULL,
	uuser	varchar(60)	default NULL,
	utime	datetime	default NULL,
	vtime	datetime	default NULL,
	puser	varchar(60)	default NULL,
	prole	varchar(60)	default NULL,
	auser	varchar(60)	default NULL,
	arole	varchar(60)	default NULL,
	mrole	varchar(60)	default NULL,
	rrole	varchar(60)	default NULL,
	mailto	varchar(255)	default NULL,
	mailtime	datetime	default NULL,
	status	varchar(10)	default NULL,
	severity	decimal(1,0)	default NULL,
	etime	datetime	default NULL,
	stime	datetime	default NULL,
	period	varchar(20)	default NULL,
	record	varchar(10)	default NULL,
	rectype	varchar(10)	default NULL,
	recprc	varchar(10)	default NULL,
	object	varchar(60)	default NULL,
	application	varchar(80)	default NULL,
	location	varchar(80)	default NULL,
	cause	varchar(60)	default NULL,
	process	varchar(80)	default NULL,
	cost	decimal(10,2)	default NULL,
	analysis varchar(255)	default NULL,
	subject	varchar(255)	default NULL,
	comment	text,
	PRIMARY KEY  (id),
	KEY idnv	(idnv,etime,utime),
	KEY idpr	(idpr,etime,utime),
	KEY idrm	(idrm,etime,utime),
	KEY idrr	(idrr,etime,utime),
	KEY record	(record,etime,utime),
	KEY object	(object,etime,utime),
	KEY doctype	(doctype,etime,utime),
	KEY auser	(auser,etime,utime),
	KEY arole	(arole,etime,utime),
	KEY application	(application,etime,utime),
	KEY location	(location,etime,utime),
	KEY process	(process,etime,utime)
);
{"'cmdbm' add stmt"
 ###########################
}
# 
ALTER TABLE cmdbm ADD COLUMN idold varchar(60) AFTER id;
ALTER TABLE cmdbm ADD COLUMN stmt varchar(80) AFTER definition;
ALTER TABLE cmdbm ADD COLUMN vtime datetime AFTER utime;
#
#
#
{"'hdesk' 2007-01-24 add 'analysis'" 
 ###########################
}
#
ALTER TABLE hdesk ADD COLUMN analysis varchar(255) AFTER doctype;
#
#
#
{"'hdesk' 2007-04-09 drop unused cols, add 'vtime','mrole','recprc'..."
 ###########################
}
#
ALTER TABLE hdesk DROP idpt, DROP idlr, DROP lslote, DROP progress, DROP aopt, DROP doctype;
ALTER TABLE hdesk ADD COLUMN vtime datetime default NULL AFTER utime;
ALTER TABLE hdesk ADD COLUMN mrole varchar(60) default NULL AFTER arole;
ALTER TABLE hdesk ADD COLUMN recprc varchar(10) default NULL AFTER record;
ALTER TABLE hdesk ADD COLUMN rectype varchar(10) default NULL AFTER record;
ALTER TABLE hdesk ADD COLUMN cause varchar(60) default NULL AFTER location;
#
{"'notes' table" 2007-05-04
 ###########################
}
# 
CREATE TABLE notes (
        id       varchar(60) primary key,
        idnv     varchar(60),  #   new version (value) pointer
        idpr     varchar(60),  #   previous record pointer
        idrm     varchar(60),  #   reply master pointer
        cuser    varchar(60),  #   creator user
        ctime    datetime,     #   created time
        uuser    varchar(60),  #   updator user
        utime    datetime,     #   updated time
        prole    varchar(60),  #   principal role
        rrole    varchar(60),  #   reader role
	mailto	 varchar(255), #   email receipients
        status   varchar(10),  #   record status
        record   varchar(10),  # - record type
        object   varchar(60),  # - object name
        doctype  varchar(60),  # - document type
        subject  varchar(255), #   subject, title
        comment  text          #   comment, text
)
;
CREATE INDEX idrm     ON notes (idrm,    utime desc,  ctime desc);
CREATE INDEX cuser    ON notes (cuser,   utime desc,  ctime desc);
CREATE INDEX uuser    ON notes (uuser,   utime desc,  ctime desc);
CREATE INDEX prole    ON notes (prole,   utime desc,  ctime desc);
CREATE INDEX rrole    ON notes (rrole,   utime desc,  ctime desc);
CREATE INDEX subject  ON notes (subject, utime desc,  ctime desc);
#
#